Close #1, add ability to upload files

plocket opened the pull request:
Closes #1. I used the files in the example folder for testing. Sorry for the unnecessary edits to wording and a bit to styles. Had trouble resisting getting those changes in. I can specifically note those if that would help you ignore them.

plocket committed:
Close #1, Allow uploading of var data (and exclusion strings)

Adds file for testing upload of exclusion strings (empty list)

plocket committed:
Improve error handling, refactor for such purpose, improve comments

and tweak wording and style. Sorry there's so much here.

niharikasingh commented:
A few comments:

1. It would be helpful to me if you just listed the other edits in this MR. That way I can map what you intended to what's in the code. Just a one sentence description for each of the changes will suffice. 
2. I noticed a minor bug. Repro steps:

- Create a JSON file with an error. I used a blank file. 
- Upload the JSON file, see error. 
- Edit the JSON file to have proper JSON. 
- Upload file again. Error doesn't disappear and upload doesn't seem to work. 

3. Very minor stylistic thing, but it would be nice to have more padding before/after this error message. 
<img width="611" alt="Screen Shot 2021-06-02 at 9 03 27 PM" src="https://user-images.githubusercontent.com/3387590/120570605-82d7b080-c3e6-11eb-814b-82daf902b447.png">

plocket commented:
Fair request. Detailed list of changes (let me know if it's sufficient):
1. Used actual `ignore_anywhere` instead of the default `ignore_anywhere` to exclude rows.
1. Added ability to upload a list of strings to cause a row to be excluded from the table.
2. Because of that, abstracted some functionality for setting values and handling row exclusion/ignoring input errors. Just now recognized that I went overboard - the user will not be able to edit the text by hand, so will have to fix that.
2. Added ability to upload a file to set variable data.
2. Because of that, abstracted handling the error for the variable data input. Apparently not working completely yet (thanks for catching that).
3. Abstracted functionality to change a textarea's size based on its contents with `fit_textarea_to_content`, though looking at it now, I'm not sure I'm totally happy with the way it's currently handled.
3. Renamed some variables as their purpose confused me.
2. Added/edited comments.
3. Adjusted styles and some text content to, I hope, make things more clear.

plocket committed:
Fix no change when reload same file. Reset val of input on upload

plocket committed:
Fix can't edit textarea for excluding rows with specific text

plocket committed:
Simplify var data error placement

plocket commented:
Hope that's been improved, @niharikasingh

niharikasingh commented on generate.js:
```diff
@@ -450,73 +454,128 @@ let update_output = function () {
+      if ( typeof new_value === `string` ) { maybe_exclusion_list = JSON.parse( new_value ); }
+      if ( !Array.isArray( maybe_exclusion_list )) {
+        // Does not yet test that every item is a string
+        let error = `Error: The data is not a JSON list of strings. This tool can only use a list of strings to exclude rows in the story table.`
```
I'm not sure that this is working as intended... This is how I tested it: 

1. Create JSON file with contents `{"bob"}` (this is an object, not an array, so should error)
2. Upload this JSON file in the box that says "Customize the list of text that will cause a row in the table to be excluded."
3. The error message I see is `Could not load "tests.feature". SyntaxError: Unexpected token } in JSON at position 6`, which is not as helpful as the text here

niharikasingh replied:
Ah nvm, my object was poorly constructed, duh. This works if the file I upload says `{"bob": "duck"}`. Nothing to see here.
